rate-limiter/src/rate_limiter.rs

```diff
@@ -0,0 +1,90 @@
+use std::{
+    pin::Pin,
+    time::{Duration, Instant},
+};
+
+use futures::Future;
+use tokio::time::Sleep;
+
+use crate::token_bucket::TokenBucket;
+
+pub struct SleepingRateLimiter {
+    rate_limiter: TokenBucket,
+    sleep: Pin<Box<Sleep>>,
+    finished: bool,
+}
+
+impl Clone for SleepingRateLimiter {
+    fn clone(&self) -> Self {
+        Self::new(self.rate_limiter.clone())
+    }
+}
+
+impl SleepingRateLimiter {
+    pub fn new(rate_limiter: TokenBucket) -> Self {
+        Self {
+            rate_limiter,
+            sleep: Box::pin(tokio::time::sleep(Duration::ZERO)),
+            finished: true,
+        }
+    }
+
+    pub fn into_inner(self) -> TokenBucket {
+        self.rate_limiter
+    }
+}
+
+impl SleepingRateLimiter {
+    fn set_sleep(&mut self, read_size: usize) -> RateLimiterTask {
+        let mut now = None;
+        let mut now_closure = || *now.get_or_insert_with(Instant::now);
+        let next_wait = self.rate_limiter.rate_limit(read_size, &mut now_closure);
+        if let Some(next_wait) = next_wait {
+            let wait_until = now_closure() + next_wait;
+            self.finished = false;
+            self.sleep.set(tokio::time::sleep_until(wait_until.into()));
+        }
+        self.current_sleep()
+    }
+
+    pub fn rate_limit(&mut self, read_size: usize) -> RateLimiterTask {
+        self.set_sleep(read_size)
+    }
+
+    pub fn current_sleep(&mut self) -> RateLimiterTask {
+        RateLimiterTask::new(&mut self.sleep, &mut self.finished)
```

Comment:
Hmmm, is this really necessary, or could this just be an `async` method calling the inner sleep? You can poll the results of async methods, right? Then you would need to just do polling at the point where you are implementing `AsyncRead` which requires explicit polling.

I might very well be wrong about this (I'm not confident in my knowledge about `async` in detail, especially with pinning), but then I would be quite interested in an explanation as to what the problem here is.

Reply:
Removed. Using some other trick instead. Now it uses `#![feature(type_alias_impl_trait)]` to retrieve the type of an async. I don't like that it's a extra feature, but like the idea that `dyn` or `BoxFuture` is not needed for async.

Reply:
Uhoh, I see, I already wrote another comment about this. >.< This is a nightly feature, right? I don't think using these is a good idea, are there no other decent solutions?

Reply:
Alternatives I know are `Pin<Box<dyn Future>>` aka `BoxFuture`, or a custom `Future` impl. I switched it to `BoxFuture`.